File: samples/node/src/connectedcarsapi.js
const axios = require('axios')
const { JwtUtils } = require('@connectedcars/jwtutils')

const _readServiceAccountData = ccServiceAccountKeyData => {
  const [ccInfo, rsa] = ccServiceAccountKeyData.split(
    '----- END CONNECTEDCARS INFO -----\n'
  )

  const ccRegex = /----- BEGIN CONNECTEDCARS INFO -----\niss: (.*)\naud: (.*)\nkid: (.*)\n/
  const [totalMatch, iss, aud, kid] = ccInfo.match(ccRegex)

  if (!(iss && aud && kid)) {
    throw new Error('Malformed service account file')
  }

  return {
    iss,
    aud,
    kid,
    rsa
  }
}

const _getToken = async (parsedServiceAccountInfo, authApiEndpoint) => {
  try {
    const unixNow = Math.floor(Date.now() / 1000)

    let jwtHeader = {
      typ: 'JWT',
      alg: 'RS256',
      kid: parsedServiceAccountInfo.kid
    }

    let jwtBody = {
      aud: parsedServiceAccountInfo.aud,
      iss: parsedServiceAccountInfo.iss,
      iat: unixNow,
      exp: unixNow + 3600
    }

    let jwt = JwtUtils.encode(parsedServiceAccountInfo.rsa, jwtHeader, jwtBody)

    const res = await axios.default.post(authApiEndpoint, { token: jwt })
    if (!res.data.token) {
      throw new Error('No token returned')
    }
    return res.data
  } catch (err) {
    throw err
  }
}

class ConnectedCarsApi {
  /**
   * Create an instance of the Connected Cars api, which can be used to call the GraphQL api. Requires specifying service account key data and an environment
   * @param {string} ccServiceAccountKeyData a string containing the Connected Cars service account data
   * @param {string} [endpoint] specify the connected cars api endpoint, default is production endpoint
   * @param {string} [authEndpoint] specify the connected cars auth endpoint, default is production endpoint
   * @throws an error if the service account data is malformed
   */
  constructor(
    ccServiceAccountKeyData,
    endpoint = 'https://api.connectedcars.io/graphql',
    authEndpoint = 'https://auth-api.connectedcars.io/auth/login/serviceAccountConverter'
  ) {
    this._API_ENDPOINT = endpoint
    this._AUTH_API_ENDPOINT = authEndpoint
    this._ccAccessToken = null
    this._parsedServiceAccountInfo = _readServiceAccountData(
      ccServiceAccountKeyData
    )
  }

  /**
   * Gets a Connected Cars access token, if you want to interact with the api without using the call() method
   * @returns {Promise<string>}
   * @throws an error if the CC auth api cannot validate your credentials
   */
  async getAccessToken() {
    const now = Date.now()
    if (
      !this._ccAccessToken ||
      this._ccAccessToken.expires < now + 5 * 60 * 1000
    ) {
      this._ccAccessToken = await _getToken(
        this._parsedServiceAccountInfo,
        this._AUTH_API_ENDPOINT
      )
    }
    return this._ccAccessToken.token
  }

  /**
   * Clears the Connected Cars token, which will force a refetch next time call() is used.
   * This is done automatically in case the token is invalid on a call()
   */
  async _clearToken() {
    this._ccAccessToken = null
  }

  /**
   * Call the Connected Cars GraphQL api
   * @param {Object} graphQLInput Should be valid graphql input, such as `query User {user(id:52163) {id firstname} }`
   * @returns {Promise<Object>} returns a graphql response such as user: {id: "52163", firstname: null}
   * @throws an error if the call to the CC api fails, or if the CC auth api cannot validate your credentials
   */
  async call(graphQLInput) {
    const bearerToken = await this.getAccessToken()

    const config = {
      headers: { Authorization: 'Bearer ' + bearerToken }
    }
    const query = {
      query: graphQLInput
    }

    try {
      const result = await axios.default.post(this._API_ENDPOINT, query, config)
      return result.data.data
    } catch (err) {
      // Retry once with a new token in case of 401
      if (err.response && err.response.status === 401) {
        await this._clearToken()
        const newBearerToken = await this.getAccessToken()
        const newConfig = {
          headers: { Authorization: 'Bearer ' + newBearerToken }
        }
        const result = await axios.default.post(
          this._API_ENDPOINT,
          graphQLInput,
          newConfig
        )
        return result.data.data
      }
      throw err
    }
  }
}

module.exports = ConnectedCarsApi

File: samples/java/full-example/connectedcars-api/src/main/java/io/connectedcars/api/ConnectedCarsApi.java
package io.connectedcars.api;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.security.GeneralSecurityException;
import java.time.Instant;
import java.util.stream.Collectors;


import com.fasterxml.jackson.databind.ObjectMapper;
import io.connectedcars.authentication.*;
import org.apache.http.HttpEntity;
import org.apache.http.HttpResponse;
import org.apache.http.client.HttpClient;
import org.apache.http.client.methods.HttpPost;
import org.apache.http.entity.StringEntity;
import org.apache.http.impl.client.HttpClients;


public class ConnectedCarsApi {
    private String endpoint;
    private String authEndpoint;
    private String organizationNamespace;

    private ServiceAccount serviceAccount;
    private CCAccessToken CCAccessToken;
    private ObjectMapper mapper = new ObjectMapper();
    private HttpClient httpclient = HttpClients.createDefault();


    /**
     * Create an instance of the Connected Cars API, which can be used to call the GraphQL API. Requires specifying service account key data and endpoints
     * @param ccServiceAccountKeyData a string containing the Connected Cars service account data
     * @param endpoint specify the connected cars api endpoint, production is: https://api.connectedcars.io/graphql
     * @param authEndpoint specify the connected cars auth endpoint, production is: https://auth-api.connectedcars.io/auth/login/serviceAccountConverter
     * @throws GeneralSecurityException
     * @throws IOException
     */
    public ConnectedCarsApi(String ccServiceAccountKeyData, String endpoint, String authEndpoint, String organizationNamespace) throws GeneralSecurityException, IOException {
        this.serviceAccount = new ServiceAccount(ccServiceAccountKeyData);
        this.endpoint = endpoint;
        this.authEndpoint = authEndpoint;
        this.organizationNamespace = organizationNamespace;
    }

    private CCAccessToken getToken() throws IOException {

        HttpPost httppost = new HttpPost(authEndpoint);

        // Wrap the token in a class for JSON
        TokenWrapper serviceAccountToken = new TokenWrapper();
        serviceAccountToken.token = this.serviceAccount.getToken();

        httppost.addHeader("content-type", "application/json");
        httppost.setEntity(new StringEntity(mapper.writeValueAsString(serviceAccountToken)));

        //Execute and get the response.
        HttpResponse response = httpclient.execute(httppost);
        if (response.getStatusLine().getStatusCode() != 200) {
            throw new RuntimeException(response.getStatusLine().getStatusCode() + " " + response.getStatusLine().getReasonPhrase());
        }

        HttpEntity entity = response.getEntity();

        if (entity != null) {
            try (InputStream inStream = entity.getContent()) {
                String result = new BufferedReader(new InputStreamReader(inStream))
                        .lines().collect(Collectors.joining("\n"));
                return mapper.readValue(result, CCAccessToken.class);
            }
        }
        throw new RuntimeException("Unknown error");
    }

    /**
     * Get an access token for the Connected Cars API
     * @return an access token for the Connected Cars API
     * @throws IOException
     */
    public String getAccessToken() throws IOException {
        long unixTime = Instant.now().getEpochSecond();
        if (this.CCAccessToken == null || this.CCAccessToken.getExpires() < unixTime + 5 * 60 * 1000) {
            this.CCAccessToken = this.getToken();
        }
        return this.CCAccessToken.getToken();
    }

    private void clearToken() {
        this.CCAccessToken = null;
    }

    /**
     * Call the Connected Cars GraphQL API
     * @param graphQLInput Should be valid graphql input, such as "query User {user(id:52163) {id firstname} }"
     * @return a graphql response such as user: {id: "52163", firstname: null}
     * @throws RuntimeException
     * @throws IOException
     */
    public String call(String graphQLInput) throws RuntimeException, IOException {
        String accessToken = this.getAccessToken();

        HttpPost httppost = new HttpPost(endpoint);

        // Wrap the query in a class for JSON
        QueryWrapper query = new QueryWrapper();
        query.query = graphQLInput;

        httppost.addHeader("content-type", "application/json");
        httppost.addHeader("Authorization","Bearer " + accessToken);
        httppost.setEntity(new StringEntity(mapper.writeValueAsString(query)));

        //Execute and get the response.
        HttpResponse response = httpclient.execute(httppost);
        if (response.getStatusLine().getStatusCode() == 401) {
            // Retry once with new token in case of 401
            this.clearToken();
            String newAccessToken = this.getAccessToken();

            HttpPost post = new HttpPost(endpoint);

            post.addHeader("content-type", "application/json");
            post.addHeader("Authorization","Bearer " + newAccessToken);
            post.addHeader("X-Organization-Namespace", this.organizationNamespace);
            post.setEntity(new StringEntity(mapper.writeValueAsString(query)));

            //Execute and get the response.
            HttpResponse secondResponse = httpclient.execute(post);
            if (secondResponse.getStatusLine().getStatusCode() != 200) {
                throw new RuntimeException(secondResponse.getStatusLine().getStatusCode() + " " + secondResponse.getStatusLine().getReasonPhrase());
            }
            HttpEntity secondEntity = secondResponse.getEntity();
            if (secondEntity != null) {
                try (InputStream inStream = secondEntity.getContent()) {
                    String result = new BufferedReader(new InputStreamReader(inStream))
                            .lines().collect(Collectors.joining("\n"));
                    return result;
                }
            }
            throw new RuntimeException("Unknown error");
        }
        HttpEntity entity = response.getEntity();

        if (entity != null) {
            try (InputStream inStream = entity.getContent()) {
                String result = new BufferedReader(new InputStreamReader(inStream))
                        .lines().collect(Collectors.joining("\n"));
                return result;
            }
        }
        throw new RuntimeException(response.getStatusLine().getStatusCode() + " " + response.getStatusLine().getReasonPhrase());
    }

    private static class TokenWrapper {
        public String token;
    }

    private static class QueryWrapper {
        public String query;
    }
}



